### PR TITLE
Unsafe Keychain Accessibility

### PR DESCRIPTION
**Pull request by benyissa**
recommendation preview : 

image

image

### CODE REVIEW

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/meta.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{
+  "risk_rating": "high",
+  "short_description": "Insecure Storage: Unsafe Keychain Accessibility vulnerability allows sensitive data stored on the device to be easily accessed by malicious actors.",
+  "references": {
+    "Apple documentation: Restricting keychain item accessibility": "https://developer.apple.com/documentation/security/keychain_services/keychain_items/restricting_keychain_item_accessibility",
+    "Secure Mobile Biometric Authentication": "https://blog.ostorlab.co/secure-mobile-biometric-authentication.html"
```
Comment (PiranhaSa) [inline MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/meta.json:6]: keychain is used in multiple things not only secure bio

Comment (amine3) [reply MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/meta.json:6]: It is ok to have this reference since it provides more details.

**Review by amine3 — APPROVED** [review]

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            kSecAttrAccessible as String: kSecAttrAccessibleAlways
+        ]
+        SecItemAdd(query as CFDictionary, nil)
+    }
+    saveSecretToKeychain()
+	```
+
+In the following examples, incomplete specification of access control flags (`SecAccessControlCreateFlags`) results in insecure keychain item storage.
+
+=== "Swift"
```
Comment (rabsondev) [inline MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md:24]: What's the difference between this example and the one below?

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
+Insecure storage occurs when sensitive data is stored without proper protection mechanisms, such as those provided by the iOS keychain. This exposes the data to potential unauthorized access, posing a significant security risk for iOS apps.
+
+In this example, the secret password is stored in the keychain with unrestricted accessibility (`kSecAttrAccessibleAlways`), leaving it vulnerable to unauthorized access.
```
Comment (elyousfi5) [inline MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md:3]:
```suggestion
In this example, the secret password is stored in the keychain with unrestricted accessibility (`kSecAttrAccessibleAlways`), leaving it vulnerable to unauthorized access:
```

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            kSecAttrAccount as String: "myAccount",
+            kSecValueData as String: secretData,
+            kSecAttrAccessible as String: kSecAttrAccessibleAlways
+        ]
+        SecItemAdd(query as CFDictionary, nil)
+    }
+    saveSecretToKeychain()
+	```
+
+In the following examples, incomplete specification of access control flags (`SecAccessControlCreateFlags`) results in insecure keychain item storage.
```
Comment (elyousfi5) [inline MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md:22]:
```suggestion
In the following examples, incomplete specification of access control flags (`SecAccessControlCreateFlags`) results in insecure keychain item storage:
```

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        let query: [String: Any] = [
+            kSecClass as String: kSecClassGenericPassword,
+            kSecAttrAccount as String: "myAccount",
+            kSecValueData as String: secretData,
+            kSecAttrAccessControl as String: access!,
+        ]
+        SecItemAdd(query as CFDictionary, nil)
+    }
+    saveSecretToKeychain()
+    ```
```
Comment (elyousfi5) [inline MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md:70]: This is duplicated

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/recommendation.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
+**1. Restrict Keychain Item Accessibility:**
+
+   Set appropriate conditions for accessing keychain items based on the device state and user preferences.
```
Comment (elyousfi5) [inline MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/recommendation.md:3]: I dont think that text with indentation will look nice in the vuln page.

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md (hunk 2 above)
Comment (benyissa) [reply MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md:24]: one is .or and other one is .and

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md (hunk 5 above)
Comment (benyissa) [reply MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md:70]: it's not one is .or and other one is .and

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/recommendation.md (hunk 6 above)
Comment (benyissa) [reply MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/recommendation.md:3]: This is a recommendation preview : 

image

I think it's fine.

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
+Insecure storage occurs when sensitive data is stored without proper protection mechanisms, such as those provided by the iOS keychain. This exposes the data to potential unauthorized access, posing a significant security risk for iOS apps.
```
Comment (3asm) [inline MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/description.md:1]: Make it direct. The platform is using keychain and the keychain is using insecure flags.

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/meta.json [hunk 8]
```diff
@@ -0,0 +1,27 @@
+{
+  "risk_rating": "high",
+  "short_description": "Insecure Storage: Unsafe Keychain Accessibility vulnerability allows sensitive data stored on the device to be easily accessed by malicious actors.",
```
Comment (3asm) [inline MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/meta.json:3]: Please use a sentence and remove the :

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/meta.json [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  "references": {
+    "Apple documentation: Restricting keychain item accessibility": "https://developer.apple.com/documentation/security/keychain_services/keychain_items/restricting_keychain_item_accessibility",
+    "Secure Mobile Biometric Authentication": "https://blog.ostorlab.co/secure-mobile-biometric-authentication.html"
+  },
+  "title": "Insecure Storage: Unsafe Keychain Accessibility",
```
Comment (3asm) [inline MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/meta.json:8]:
```suggestion
  "title": "Insecure Keychain Storage",
```
Simple is better than complicated.
Complicated is better than complex. ;)

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/recommendation.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
+**1. Restrict Keychain Item Accessibility:**
```
Comment (3asm) [inline MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/recommendation.md:1]: Is accessibility used in Apple documentation? Accessibility is typically used for users with disabilities.

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/recommendation.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+   Set appropriate conditions for accessing keychain items based on the device state and user preferences.
+
+   - For sensitive data that should only be accessed when the device is unlocked, use `kSecAttrAccessibleWhenUnlocked`.
+   - If access to the keychain item is required after the user unlocks the device for the first time, use `kSecAttrAccessibleAfterFirstUnlock`.
+   - Avoid using `kSecAttrAccessibleAlways`, as it provides unrestricted access to the keychain item regardless of the device's locked state, which can compromise security.
```
Comment (3asm) [inline MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/recommendation.md:7]: How about biometric flags?

Comment (3asm) [reply MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/recommendation.md:7]: Would be nice to have a table instead of what flag to use and when.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/recommendation.md (hunk 10 above)
Comment (benyissa) [reply MOBILE_CLIENT/IOS/_HIGH/UNSAFE_KEYCHAIN_ACCESSIBILITY/recommendation.md:1]: Yes.
This is a reference: https://developer.apple.com/documentation/security/keychain_services/keychain_items/restricting_keychain_item_accessibility?language=objc

**Review by 3asm — APPROVED** [review]

**Pull request merged**